File: node/test/integration.test.ts
import { readFileSync } from "node:fs";
import { createServer, type Server } from "node:http";
import type { AddressInfo } from "node:net";
import { fileURLToPath } from "node:url";
import { afterAll, beforeAll, describe, expect, it } from "vitest";
import { OpenBankingClient } from "../src/index.js";

const fixtures = fileURLToPath(new URL("../../fixtures/", import.meta.url));
const readJson = (name: string) => JSON.parse(readFileSync(fixtures + name, "utf8"));

const API_KEY = readJson("credentials.json").apiKey as string;
const PRIVATE_KEY = readJson("keypair.json").privateKeyPkcs8B64 as string;

let server: Server;
let baseUrl: string;
let lastSyncBody: unknown = null;

function send(res: import("node:http").ServerResponse, status: number, body: unknown): void {
  res.writeHead(status, { "Content-Type": "application/json" });
  res.end(JSON.stringify(body));
}

beforeAll(async () => {
  server = createServer((req, res) => {
    // Enforce the API key — 401 otherwise.
    if (req.headers["x-api-key"] !== API_KEY) {
      send(res, 401, { error: "unauthorized" });
      return;
    }

    const url = new URL(req.url ?? "/", "http://localhost");
    const path = url.pathname;

    if (req.method === "GET" && path === "/api/accounts") {
      send(res, 200, readJson("api/accounts.json"));
      return;
    }
    if (req.method === "GET" && /^\/api\/accounts\/[^/]+\/transactions$/.test(path)) {
      send(res, 200, readJson("api/transactions.json"));
      return;
    }
    if (req.method === "GET" && path === "/api/connections") {
      send(res, 200, readJson("api/connections.json"));
      return;
    }
    if (req.method === "POST" && /^\/api\/accounts\/[^/]+\/sync$/.test(path)) {
      let raw = "";
      req.on("data", (c) => (raw += c));
      req.on("end", () => {
        lastSyncBody = JSON.parse(raw || "{}");
        send(res, 200, readJson("api/sync.json"));
      });
      return;
    }
    if (req.method === "POST" && path === "/api/sync") {
      let raw = "";
      req.on("data", (c) => (raw += c));
      req.on("end", () => {
        lastSyncBody = JSON.parse(raw || "{}");
        send(res, 200, readJson("api/sync-all.json"));
      });
      return;
    }

    send(res, 404, { error: "not found" });
  });

  await new Promise<void>((resolve) => server.listen(0, "127.0.0.1", resolve));
  const port = (server.address() as AddressInfo).port;
  baseUrl = `http://127.0.0.1:${port}`;
});

afterAll(async () => {
  await new Promise<void>((resolve) => server.close(() => resolve()));
});

function client(apiKey = API_KEY): OpenBankingClient {
  return new OpenBankingClient({ apiBaseUrl: baseUrl, apiKey, privateKeyPkcs8: PRIVATE_KEY });
}

describe("integration against a mock API", () => {
  it("getAccounts decrypts iban/ownerName/displayName and 3 balances", async () => {
    const accounts = await client().getAccounts();
    expect(accounts).toHaveLength(1);
    const a = accounts[0]!;

    expect(a.iban).toBe("DK6466952001724927");
    expect(a.ownerName).toBe("Tatic ApS");
    expect(a.displayName).toBe("Drift");
    expect(a.balances).toHaveLength(3);

    const itbd = a.balances.find((b) => b.type === "ITBD");
    const itav = a.balances.find((b) => b.type === "ITAV");
    expect(itbd?.amount).toBe("828.13");
    expect(itav?.amount).toBe("633.90");
  });

  it("getTransactions decrypts amount and creditor", async () => {
    const page = await client().getTransactions("11111111-1111-4111-8111-111111111111", {
      limit: 50,
    });
    expect(page.total).toBe(1);
    const t = page.items[0]!;
    expect(t.amount).toBe("194.23");
    expect(t.creditorName).toBe("One.com");
  });

  it("getConnections works", async () => {
    const conns = await client().getConnections();
    expect(conns).toHaveLength(1);
    expect(conns[0]!.aspspName).toBe("Lunar");
    expect(conns[0]!.status).toBe("Active");
  });

  it("sync POSTs a body containing the decrypted uid", async () => {
    lastSyncBody = null;
    const result = await client().sync("11111111-1111-4111-8111-111111111111");
    expect(result.totalFetched).toBe(1);
    expect(lastSyncBody).toEqual({ uid: "c5d93aa7-5e23-4da0-ba88-42b9a584492c" });
  });

  it("syncAll POSTs items with the decrypted uid", async () => {
    lastSyncBody = null;
    const result = await client().syncAll();
    expect(result.accounts).toBe(1);
    expect(lastSyncBody).toEqual({
      items: [
        {
          accountId: "11111111-1111-4111-8111-111111111111",
          uid: "c5d93aa7-5e23-4da0-ba88-42b9a584492c",
        },
      ],
    });
  });

  it("a wrong API key throws", async () => {
    await expect(client("wrong-key").getAccounts()).rejects.toThrow();
  });

  it("a wrong private key throws on decryption", async () => {
    const bad = new OpenBankingClient({
      apiBaseUrl: baseUrl,
      apiKey: API_KEY,
      // A different valid P-256 PKCS#8 key — derives the wrong shared secret.
      privateKeyPkcs8:
        "MIGHAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBG0wawIBAQQgdB9vjCwdrF1FckSNDHrw8M7PNNkpUB/tAK/EgAbZWvihRANCAASgg8XOKlU9VeYCee9+tQKtDSkFze10CRTA4b2gGKDlHIFw+QTf1AkjAjLfWqCJ4BctUqQtAYs0v0Y90Bw1wsBF",
    });
    await expect(bad.getAccounts()).rejects.toThrow();
  });
});

File: java/src/test/java/io/openbanking/IntegrationTest.java
package io.openbanking;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;
import java.io.IOException;
import java.math.BigDecimal;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

/** Integration tests against a mock API served from the shared fixtures. */
class IntegrationTest {

  private static final Path FIXTURES = Path.of("..", "fixtures");
  private static final ObjectMapper MAPPER = new ObjectMapper();

  private HttpServer server;
  private String baseUrl;
  private String apiKey;
  private String privateKey;
  private Map<String, Object> lastSyncBody;
  private String lastUserAgent;

  private static byte[] fixture(String name) throws IOException {
    return Files.readAllBytes(FIXTURES.resolve(name));
  }

  @SuppressWarnings("unchecked")
  private static Map<String, Object> readJson(String name) throws IOException {
    return MAPPER.readValue(fixture(name), Map.class);
  }

  @BeforeEach
  void setUp() throws Exception {
    apiKey = (String) readJson("credentials.json").get("apiKey");
    privateKey = (String) readJson("keypair.json").get("privateKeyPkcs8B64");

    server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
    server.createContext("/", this::handle);
    server.start();
    baseUrl = "http://127.0.0.1:" + server.getAddress().getPort();
  }

  @AfterEach
  void tearDown() {
    server.stop(0);
  }

  private void handle(HttpExchange ex) throws IOException {
    try {
      if (!apiKey.equals(ex.getRequestHeaders().getFirst("X-Api-Key"))) {
        respond(ex, 401, "{\"error\":\"unauthorized\"}".getBytes(StandardCharsets.UTF_8));
        return;
      }
      lastUserAgent = ex.getRequestHeaders().getFirst("User-Agent");
      String method = ex.getRequestMethod();
      String path = ex.getRequestURI().getPath();

      if (method.equals("GET") && path.equals("/api/accounts")) {
        respond(ex, 200, fixture("api/accounts.json"));
      } else if (method.equals("GET") && path.matches("^/api/accounts/[^/]+/transactions$")) {
        respond(ex, 200, fixture("api/transactions.json"));
      } else if (method.equals("GET") && path.equals("/api/connections")) {
        respond(ex, 200, fixture("api/connections.json"));
      } else if (method.equals("POST") && path.matches("^/api/accounts/[^/]+/sync$")) {
        captureBody(ex);
        respond(ex, 200, fixture("api/sync.json"));
      } else if (method.equals("POST") && path.equals("/api/sync")) {
        captureBody(ex);
        respond(ex, 200, fixture("api/sync-all.json"));
      } else {
        respond(ex, 404, "{\"error\":\"not found\"}".getBytes(StandardCharsets.UTF_8));
      }
    } catch (Exception e) {
      respond(ex, 500, ("{\"error\":\"" + e.getMessage() + "\"}").getBytes(StandardCharsets.UTF_8));
    }
  }

  @SuppressWarnings("unchecked")
  private void captureBody(HttpExchange ex) throws IOException {
    byte[] body = ex.getRequestBody().readAllBytes();
    lastSyncBody = MAPPER.readValue(body, Map.class);
  }

  private static void respond(HttpExchange ex, int status, byte[] body) throws IOException {
    ex.getResponseHeaders().set("Content-Type", "application/json");
    ex.sendResponseHeaders(status, body.length);
    ex.getResponseBody().write(body);
    ex.close();
  }

  private OpenBankingClient client() {
    return new OpenBankingClient(baseUrl, apiKey, privateKey, null);
  }

  @Test
  void getAccountsDecryptsFields() {
    List<Account> accounts = client().getAccounts();
    assertEquals(1, accounts.size());
    Account a = accounts.get(0);
    assertEquals("DK6466952001724927", a.iban());
    assertEquals("Tatic ApS", a.ownerName());
    assertEquals("Drift", a.displayName());
    assertEquals(3, a.balances().size());

    Map<String, BigDecimal> byType = new java.util.HashMap<>();
    for (Balance b : a.balances()) {
      byType.put(b.type(), b.amount());
    }
    assertEquals(0, new BigDecimal("828.13").compareTo(byType.get("ITBD")));
    assertEquals(0, new BigDecimal("633.90").compareTo(byType.get("ITAV")));
  }

  @Test
  void getTransactionsDecryptsAmountAndCreditor() {
    TransactionPage page =
        client()
            .getTransactions(
                "11111111-1111-4111-8111-111111111111", TransactionQuery.none().withLimit(50));
    assertEquals(1, page.total());
    Transaction t = page.items().get(0);
    assertEquals(0, new BigDecimal("194.23").compareTo(t.amount()));
    assertEquals("One.com", t.creditorName());
  }

  @Test
  void getConnectionsWorks() {
    List<Connection> conns = client().getConnections();
    assertEquals(1, conns.size());
    assertEquals("Lunar", conns.get(0).aspspName());
    assertEquals("Active", conns.get(0).status());
  }

  @Test
  void syncPostsDecryptedUid() {
    SyncResult result = client().sync("11111111-1111-4111-8111-111111111111");
    assertEquals(1, result.totalFetched());
    assertNotNull(lastSyncBody);
    assertEquals("c5d93aa7-5e23-4da0-ba88-42b9a584492c", lastSyncBody.get("uid"));
  }

  @Test
  @SuppressWarnings("unchecked")
  void syncAllPostsItemsWithDecryptedUid() {
    SyncAllResult result = client().syncAll();
    assertEquals(1, result.accounts());
    assertNotNull(lastSyncBody);
    List<Map<String, Object>> items = (List<Map<String, Object>>) lastSyncBody.get("items");
    assertEquals(1, items.size());
    assertEquals("11111111-1111-4111-8111-111111111111", items.get(0).get("accountId"));
    assertEquals("c5d93aa7-5e23-4da0-ba88-42b9a584492c", items.get(0).get("uid"));
  }

  @Test
  void sendsUserAgentHeader() {
    client().getAccounts();
    assertNotNull(lastUserAgent);
    assertTrue(
        lastUserAgent.startsWith("open-banking-io/java/"),
        "unexpected User-Agent: " + lastUserAgent);
  }

  @Test
  void aWrongApiKeyThrows() {
    OpenBankingClient bad = new OpenBankingClient(baseUrl, "wrong-key", privateKey, null);
    assertThrows(OpenBankingException.class, bad::getAccounts);
  }

  @Test
  void aWrongPrivateKeyThrowsOnDecryption() {
    String badKey =
        "MIGHAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBG0wawIBAQQgdB9vjCwdrF1FckSNDHrw8M7PNNkpUB/tAK/EgAbZWvihRANCAASgg8XOKlU9VeYCee9+tQKtDSkFze10CRTA4b2gGKDlHIFw+QTf1AkjAjLfWqCJ4BctUqQtAYs0v0Y90Bw1wsBF";
    OpenBankingClient bad = new OpenBankingClient(baseUrl, apiKey, badKey, null);
    assertThrows(OpenBankingException.class, bad::getAccounts);
  }
}
